homeassistant/components/songpal/media_player.py

```diff
@@ -85,6 +91,11 @@ async def async_setup_entry(
         {vol.Required(PARAM_NAME): cv.string, vol.Required(PARAM_VALUE): cv.string},
         "async_set_sound_setting",
     )
+    platform.async_register_entity_service(
+        SET_SPEAKER_SETTING,
+        {vol.Required(PARAM_NAME): cv.string, vol.Required(PARAM_VALUE): cv.string},
```

Comment:
Would it be possible to enumerate all the settings and get their current state and create entities instead of a custom service?

Reply:
Yes, that is possible. Is that generally the preference? It would be more user friendly.

It looks like it will return 27 speakers and their settings, even if you don't use them all. I for example only have 8 speakers in use.

I would say this is probably less useful for people, mine isn't a unique use case - I use the 360 Spatial Sound Mapping, but I have different settings based on the input used. My receiver has profiles, but there doesn't appear to be a way to select a profile via API, so I just update each speaker.

Changing set_sound_setting to get the current state with associated entities would make more sense to me, as it would be more useful to users.

Reply:
Can we distinguish the speakers that are in use?

Reply:
I had assumed no, but looking at the API response it is possible. It indicates if the speaker is as a candidate (assuming speaker configuration ie 7.1) and if you are using it/have it connected. Eg: 7.1 configuration has different patterns, so 11 speakers are available, but only 8 would be in use. So some options. The Sony 'Music Center' app shows all 11 for me, with the 3 I don't use greyed out.

That makes the option more viable. Not sure what a soundbar looks like for speaker settings, but it would have been ugly to create 27 entities for speaker settings.

Reply:
Maybe create one Home Assistant device per speaker and connect the settings entities to the corresponding device? Mark the entities as entity category config.

Reply:
There is an unfinished PR #63145 which aimed to add separate entities for settings, but it was abandoned. Maybe it's worth checking out if reviving it would make sense?

Reply:
Wow, that is a big change but would improve the integration. I will close this PR and when I get some time look at PR #63145. Thanks.

Reply:
Yeah, maybe it was a bit too much in a one go indeed.. It's probably better to make this platform-by-platform to speed up the review cycles. I'm not sure if @Flameeyes is still interested on working to get these changes in, but as my old soundbar seems to be still working, I can help you with some testing & code reviews as needed.

Reply:
Yeah, I would break it up into many smaller changes, so that it can be chipped away at. My biggest issue is that I will now need to learn the ins and outs of Hass, instead of a small copy of something that is already there :)

I also don't really need this as I can already do it another way, but could be useful for others. I do like making things better, and more useful for people :)

I don't know, that PR is almost two years old.